### PR TITLE
Add anticipatedConcurrentIncoming[Bi/Uni]directionalStreams attributes & constructor arg

### LINKED ISSUE #544: Improve Stream Performance by Allowing Customizable Concurrency Limits in WebTransport
When using WebTransport to send multiple streams simultaneously, there appears to be a performance bottleneck. Specifically, sending 10,000 streams concurrently results in a delay of approximately 5 seconds for the streams to arrive, which is far from the expected near-instantaneous delivery.

Proposed Enhancement:
To address this performance issue, it would be beneficial to add a set_max_concurrent_uni_streams function. This would allow developers to customize the maximum number of unidirectional streams that can be opened concurrently, providing them with control over the system's limitations and potentially enhancing performance.

Next Steps:
Implement the set_max_concurrent_uni_streams function in the WebTransport API.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes #544 (for discussion).

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/webtransport/pull/574.html" title="Last updated on Feb 27, 2024, 10:50 PM UTC (701e688)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webtransport/574/607bfa8...jan-ivar:701e688.html" title="Last updated on Feb 27, 2024, 10:50 PM UTC (701e688)">Diff</a>

### CODE REVIEW

**Comment by jan-ivar** [comment]
Meeting:
- add a setter

File: index.bs [hunk 1]
```diff
@@ -954,6 +975,34 @@ these steps.
...
+   If not null, the user agent SHOULD attempt to reduce future round-trips by taking
+   {{[[AnticipatedConcurrentIncomingBiStreams]]}} into consideration in its
+   [=implementation-defined=] calculation of cumulative values for MAX_STREAM frames
+   it sends to the server.
+   [[!QUIC]][Section 19.11](https://www.rfc-editor.org/rfc/rfc9000#section-19.11)
+
+   The getter steps are to return [=this=]'s {{[[AnticipatedConcurrentIncomingBiStreams]]}}.
+
+   The setter steps, given |value|, are to set [=this=]'s
+   {{[[AnticipatedConcurrentIncomingBiStreams]]}} to |value|.
```
Comment (wilaw) [inline index.bs:1004]: Do we need to add some language to indicate that the user-agent is not guaranteed to provide the number of streams requested by the API?

**Comment by jan-ivar** [comment]
Meeting:
- add ref to HTTP/2
- long but plain name (bikeshed welcome!)
- remove reference to MAX_FRAMES
- present for next meeting for merging

File: index.bs (hunk 1 above)
Comment (jan-ivar) [reply index.bs]: Isn't that captured by the SHOULD?

File: index.bs [hunk 2]
```diff
@@ -954,6 +975,34 @@ these steps.
...
    [=connections=], otherwise false.
 
+: <dfn for="WebTransport" attribute>anticipatedConcurrentIncomingUnidirectionalStreams</dfn>
+:: Optionally lets an application specify the number of concurrently open
+   [=incoming unidirectional=] streams it anticipates the server creating.
+   If not null, the user agent SHOULD attempt to reduce future round-trips by taking
+   {{[[AnticipatedConcurrentIncomingUnidirectionalStreams]]}} into consideration in its
+   [=implementation-defined=] calculation of cumulative values for MAX_STREAM frames
+   it sends to the server.
+   [[!QUIC]][Section 19.11](https://www.rfc-editor.org/rfc/rfc9000#section-19.11)
```
Comment (jan-ivar) [inline index.bs:985]:
```suggestion
   negotiations with the server.
```

File: index.bs [hunk 3]
```diff
@@ -954,6 +975,34 @@ these steps.
...
+   {{[[AnticipatedConcurrentIncomingUnidirectionalStreams]]}} to |value|.
+
+: <dfn for="WebTransport" attribute>anticipatedConcurrentIncomingBidirectionalStreams</dfn>
+:: Optionally lets an application specify the number of concurrently open
+   [=bidirectional=] streams it anticipates the server creating.
+   If not null, the user agent SHOULD attempt to reduce future round-trips by taking
+   {{[[AnticipatedConcurrentIncomingBidirectionalStreams]]}} into consideration in its
+   [=implementation-defined=] calculation of cumulative values for MAX_STREAM frames
+   it sends to the server.
+   [[!QUIC]][Section 19.11](https://www.rfc-editor.org/rfc/rfc9000#section-19.11)
```
Comment (jan-ivar) [inline index.bs:999]:
```suggestion
   negotiations with the server.
```

File: index.bs [hunk 4]
```diff
@@ -1267,6 +1314,22 @@ that determine how the [=WebTransport session=] is established and used.
      </p>
    </div>
 
+: <dfn for="WebTransportOptions" dict-member>anticipatedConcurrentIncomingUnidirectionalStreams</dfn>
+:: Optionally lets an application specify the number of concurrently open
+   [=incoming unidirectional=] streams it anticipates the server creating.
+   If not null, the user agent SHOULD attempt to reduce round-trips by taking
+   {{[[AnticipatedConcurrentIncomingUnidirectionalStreams]]}} into consideration in its
+   negotiations with the server.
+   [[!QUIC]][Section 19.11](https://www.rfc-editor.org/rfc/rfc9000#section-19.11)
```
Comment (jan-ivar) [inline index.bs:1323]:
```suggestion
```

File: index.bs [hunk 5]
```diff
@@ -1267,6 +1314,21 @@ that determine how the [=WebTransport session=] is established and used.
...
+   {{[[AnticipatedConcurrentIncomingUnidirectionalStreams]]}} into consideration in its
+   negotiations with the server.
+
+: <dfn for="WebTransportOptions" dict-member>anticipatedConcurrentIncomingBidirectionalStreams</dfn>
+:: Optionally lets an application specify the number of concurrently open
+   [=bidirectional=] streams it anticipates a server creating.
+   If not null, the user agent SHOULD attempt to reduce round-trips by taking
+   {{[[AnticipatedConcurrentIncomingBidirectionalStreams]]}} into consideration in its
+   negotiations with the server.
+   [[!QUIC]][Section 19.11](https://www.rfc-editor.org/rfc/rfc9000#section-19.11)
```
Comment (jan-ivar) [inline index.bs:1330]:
```suggestion
```

File: index.bs [hunk 6]
```diff
@@ -954,6 +975,30 @@ these steps.
...
+   [=bidirectional=] streams it anticipates the server creating.
+   If not null, the user agent SHOULD attempt to reduce future round-trips by taking
+   {{[[AnticipatedConcurrentIncomingBidirectionalStreams]]}} into consideration in its
+   negotiations with the server.
+
+   The getter steps are to return [=this=]'s {{[[AnticipatedConcurrentIncomingBidirectionalStreams]]}}.
+
+   The setter steps, given |value|, are to set [=this=]'s
+   {{[[AnticipatedConcurrentIncomingBidirectionalStreams]]}} to |value|.
+
```
Comment (jan-ivar) [inline index.bs:1001]:
```suggestion

   Note: Setting {{anticipatedConcurrentIncomingUnidirectionalStreams}} or
   {{anticipatedConcurrentIncomingBidirectionalStreams}} does not guarantee
   the application will receive the number of streams it anticipates.
```

File: index.bs [hunk 7]
```diff
@@ -954,6 +975,33 @@ these steps.
...
+   Note: Setting {{anticipatedConcurrentIncomingUnidirectionalStreams}} or
+   {{anticipatedConcurrentIncomingBidirectionalStreams}} does not guarantee
+   the application will receive the number of streams it anticipates.
```
Comment (jan-ivar) [inline index.bs:1004]:
```suggestion
   the application will receive the number of streams it anticipates.

```

File: index.bs [hunk 8]
```diff
@@ -954,6 +975,34 @@ these steps.
...
```
Comment (jan-ivar) [inline index.bs:1004]:
```suggestion
Note: Setting {{anticipatedConcurrentIncomingUnidirectionalStreams}} or
{{anticipatedConcurrentIncomingBidirectionalStreams}} does not guarantee
the application will receive the number of streams it anticipates.
```

**Pull request merged**